scripts/glean.py

```diff
@@ -0,0 +1,221 @@
+    PINGS_URL_TEMPLATE = PROBE_INFO_BASE_URL + "/glean/{}/pings"
+
+    METRICS_URL_TEMPLATE = PROBE_INFO_BASE_URL + "/glean/{}/metrics"
+    PING_URL_TEMPLATE = PROBE_INFO_BASE_URL + "/glean/{}/pings"
+    DEPENDENCIES_URL_TEMPLATE = PROBE_INFO_BASE_URL + "/glean/{}/dependencies"
+
+    DEFAULT_DEPENDENCIES = ["glean"]
+
+    def __init__(self, repo, **kwargs):
+        self.repo = repo
+        self.repo_name = repo["name"]
+        self.app_id = repo["app_id"]
+
+    @staticmethod
+    def get_repos() -> List[dict]:
+        return _cache.get_json(GleanApp.REPOS_URL)
+
+    @staticmethod
+    def get_apps() -> List[GleanApp]:
+        """
+        Get all non-library Glean repositories
+        """
+        repo_data = GleanApp.get_repos()
+        # filter out repos that are just libraries, not applications
+        repos = list(
+            filter(lambda r: "library_names" not in r and r["app_id"] != "glean", repo_data)
+        )
+        return [GleanApp(repo) for repo in repos]
+
+    def get_dependencies(self):
+        # Get all of the library dependencies for the application that
+        # are also known about in the repositories file.
+
+        # The dependencies are specified using library names, but we need to
+        # map those back to the name of the repository in the repository file.
+        try:
+            dependencies = _cache.get_json(self.DEPENDENCIES_URL_TEMPLATE.format(self.repo_name))
+        except requests.HTTPError:
+            logging.info(f"For {self.repo_name}, using default Glean dependencies")
+            return self.DEFAULT_DEPENDENCIES
+
+        dependency_library_names = list(dependencies.keys())
+
+        repos_by_dependency_name = {}
+        for repo in self.get_repos():
+            for library_name in repo.get("library_names", []):
+                repos_by_dependency_name[library_name] = repo["name"]
+
+        dependencies = []
+        for name in dependency_library_names:
+            if name in repos_by_dependency_name:
+                dependencies.append(repos_by_dependency_name[name])
+
+        if len(dependencies) == 0:
+            logging.info(f"For {self.repo_name}, using default Glean dependencies")
+            return self.DEFAULT_DEPENDENCIES
+
+        logging.info(f"For {self.repo_name}, found Glean dependencies: {dependencies}")
+        return dependencies
+
+    def get_metrics(self) -> List[GleanMetric]:
+        data = _cache.get_json(GleanApp.METRICS_URL_TEMPLATE.format(self.repo_name))
+        metrics = list(data.items())
+
+        for dependency in self.get_dependencies():
+            dependency_metrics = _cache.get_json(GleanApp.METRICS_URL_TEMPLATE.format(dependency))
+            metrics += list(dependency_metrics.items())
+
+        ping_names = set(self._get_ping_data().keys())
+
+        processed = []
+        for _id, defn in metrics:
+            metric = GleanMetric(_id, defn, ping_names=ping_names)
+            processed.append(metric)
```

Comment:
I want to note https://github.com/mozilla/mozilla-schema-generator/issues/118 where some custom logic was added for schema compatibility where a field was accidentally added with the wrong type. This is specific to schema generation in mozilla-pipeline-schemas where removing a column is not done, it shouldn't matter outside of the module.

Reply:
Right, I deleted that part of the code but the structure here still reflects it. I'll update the code. If/when we use this code in mozilla-schema-generator, we can apply the transform on the output of this function.